### PR TITLE
[install] Update install

### PR DESCRIPTION
**Pull request by smousy**
- Refactor install command
- Fix bug with installation in folder component_name/component_name
- Add logic for client(web) installation
- Move same actions to utils
- Add exceptions to validate component method
- Add logic for recursive processing repositories

### CODE REVIEW

File: src/manager/commands/command_init/sc_component_manager_command_init.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@ ExecutionResult ScComponentManagerCommandInit::Execute(
     ScMemoryContext * context,
     CommandParameters const & commandParameters)
 {
-  std::pair<std::set<std::string>, std::set<std::string>> repositoryItems;
+  std::pair<std::set<std::string>, std::set<std::string>> repositoryItems; //its never updated
```
Comment (SHtress) [inline src/manager/commands/command_init/sc_component_manager_command_init.cpp:18]: Then remove it.

File: kb/specifications.scs [hunk 2]
```diff
@@ -22,9 +22,57 @@ sc_component_manager_repository
...
+           	=> nrel_alternative_addresses:
+           	...
+           	(*
+           		<- sc_node_tuple;;
+           		-> rrel_1:
+           			...
+           			(*
+           				-> [https://drive.google.com/file/d/1eUxPBd3VKNFSDEgolz11EQdEmBJMFYIk/view?usp=sharing]
+           					(*
+           						<- concept_github_url;;
```
Comment (SHtress) [inline kb/specifications.scs:71]: Change to google_drive_url.

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 3]
```diff
@@ -11,51 +11,91 @@
...
+}
 
-    SC_LOG_DEBUG("Validating component \"" + componentToInstallIdentifier);
-    if (!ValidateComponent(context, componentAddr))
+
+ExecutionResult ScComponentManagerCommandInstall::Execute(
+        ScMemoryContext *context,
+        CommandParameters const &commandParameters)
+{
+    ExecutionResult executionResult; //TODO: result doesn't change
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:73]: We can discuss this later and change it in all commands.

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 4]
```diff
@@ -134,93 +141,51 @@ bool ScComponentManagerCommandInstall::ValidateComponent(ScMemoryContext * conte
...
+    for (ScAddr const &componentDependency: componentDependencies)
     {
-      SC_LOG_ERROR("Dependency \"" + dependencyIdtf + "\" is not installed");
-      return dependencyResult;
+        std::string dependencyIdtf = context->HelperGetSystemIdtf(componentDependency);
+        SC_LOG_INFO("ScComponentManager: Install dependency \"" + dependencyIdtf + "\"");
+        CommandParameters dependencyParameters = {{PARAMETER_NAME, {dependencyIdtf}}};
+        ExecutionResult dependencyResult = Execute(context, dependencyParameters);
+
+        // Return empty if you couldn't install one from all dependencies why?
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:158]: It is more complex process, will discuss later.

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 5]
```diff
@@ -134,93 +141,51 @@ bool ScComponentManagerCommandInstall::ValidateComponent(ScMemoryContext * conte
...
-    DIR * dir;
-    struct dirent * diread;
-    componentDirName += componentAddressContent.substr(componentDirNameIndex);
-    if ((dir = opendir(componentDirName.c_str())) != nullptr)
-    {
-      while ((diread = readdir(dir)) != nullptr)
-      {
-        std::string filename = diread->d_name;
-        if (filename.rfind(".scs") != std::string::npos)
+        std::string componentDirName = componentUtils::InstallUtils::GetComponentDirName(context, componentAddr,
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:184]: Can we reuse downloader from init command?

File: src/manager/utils/sc_component_utils.cpp [hunk 6]
```diff
@@ -106,101 +106,207 @@ ScAddr SearchUtils::GetComponentInstallationMethod(ScMemoryContext * context, Sc
...
+
+        //TODO: bug: in kb folder there are 2 specifications for one rep, so all links duplicated
+        std::string script;
+        while (installScriptsIterator->Next())
+        {
+            const ScAddr &scriptAddrs = installScriptsIterator->Get(2);
+            context->GetLinkContent(scriptAddrs, script);
+            SC_LOG_DEBUG("ScComponentManager: Install script found:" + script);
+        }
+        return {};
```
Comment (SHtress) [inline src/manager/utils/sc_component_utils.cpp:232]: It always return empty vector?

File: src/manager/utils/sc_component_utils.cpp [hunk 7]
```diff
@@ -106,101 +106,207 @@ ScAddr SearchUtils::GetComponentInstallationMethod(ScMemoryContext * context, Sc
...
+            closedir(dir);
+        }
+        return result;
     }
-    closedir(dir);
-  }
-}
 
-}  // namespace componentUtils
+}  // namespace componentUtils
```
Comment (SHtress) [inline src/manager/utils/sc_component_utils.cpp:312]: Don't forget about newline

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 8]
```diff
@@ -11,51 +11,91 @@
...
+{
+    std::vector <std::string> scripts = componentUtils::InstallUtils::GetInstallScripts(context, componentAddr);
+    for (auto script: scripts)
+    {
+        std::string componentDirName = componentUtils::InstallUtils::GetComponentDirName(context, componentAddr,
+                                                                                         m_specificationsPath);
+        sc_fs_mkdirs(componentDirName.c_str());
+        ScExec exec{{"cd", componentDirName, "&&", script}};
+    }
+}
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:66]: This is gonna work only if component has install scripts as it installation method, but we have components which are just knowledge base fragments in .scs files, let's discuss how we can make universal solution.

**Review by SHtress — COMMENTED** [review]
It would be great to discuss some moments, for other i will create issues and tasks. Also there is probably some problem with formatting, let's discuss it too.

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp (hunk 5 above)
Comment (smousy) [reply src/manager/commands/command_install/sc_component_manager_command_install.cpp]: need your assistance here

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 9]
```diff
@@ -11,51 +11,93 @@
...
-    // TODO: Implement install all components method
-    SC_LOG_INFO("No identifier provided, installing all to install components");
+    ScAddrVector availableComponents;
+    for (std::string const &componentToInstallIdentifier: componentsToInstall)
+    {
+        ScAddr componentAddr = context->HelperFindBySystemIdtf(componentToInstallIdentifier);
 
-    return result;
-  }
+        SC_LOG_DEBUG("Validating component \"" + componentToInstallIdentifier);
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:34]: Don't forget to close quotation.
```suggestion
        SC_LOG_DEBUG("Validating component \"" + componentToInstallIdentifier + "\"");
```

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 10]
```diff
@@ -134,93 +143,45 @@ bool ScComponentManagerCommandInstall::ValidateComponent(ScMemoryContext * conte
...
-        {
-          loader.loadScsFile(*context, componentDirName + "/" + filename);
-        }
-      }
-      closedir(dir);
+        SC_LOG_WARNING("Not all files are loaded from" + componentDirName);
     }
-  }
-}
+}
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:187]: Add newline.

File: src/manager/commands/keynodes/ScComponentManagerKeynodes.hpp [hunk 11]
```diff
@@ -66,6 +66,9 @@ class ScComponentManagerKeynodes : public ScObject
 
   SC_PROPERTY(Keynode("nrel_repository_address"), ForceCreate(ScType::NodeConstNoRole))
   static ScAddr nrel_repository_address;
+
+  SC_PROPERTY(Keynode("nrel_install_scripts"), ForceCreate(ScType::NodeConstNoRole))
```
Comment (SHtress) [inline src/manager/commands/keynodes/ScComponentManagerKeynodes.hpp:70]: Maybe better use singular "nrel_installation_script"?

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 12]
```diff
@@ -134,93 +143,45 @@ bool ScComponentManagerCommandInstall::ValidateComponent(ScMemoryContext * conte
...
-      while ((diread = readdir(dir)) != nullptr)
-      {
-        std::string filename = diread->d_name;
-        if (filename.rfind(".scs") != std::string::npos)
-        {
-          loader.loadScsFile(*context, componentDirName + "/" + filename);
-        }
-      }
-      closedir(dir);
+        SC_LOG_WARNING("Not all files are loaded from" + componentDirName);
```
Comment (SHtress) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:185]: Need to discuss.

**Review by SHtress — APPROVED** [review]

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp (hunk 12 above)
Comment (smousy) [reply src/manager/commands/command_install/sc_component_manager_command_install.cpp:185]: comment it

**Pull request merged**